indico/modules/events/registration/controllers/management/reglists.py

```diff
@@ -789,3 +791,70 @@ def _process(self):
             if attachments_for_registration:
                 attachments[registration.id] = attachments_for_registration
         return self._generate_zip_file(attachments, name_prefix='attachments', name_suffix=self.regform.id)
+
+
+class RHManageReceiptBase(RHManageRegistrationBase):
+    normalize_url_spec = {
+        'locators': {
+            lambda self: self.receipt_file
+        }
+    }
+
+    def _process_args(self):
+        RHManageRegistrationBase._process_args(self)
+        self.receipt_file = (ReceiptFile.query
+                             .with_parent(self.registration)
+                             .filter_by(file_id=request.view_args['file_id'])
+                             .first_or_404())
+
+    def _render_receipts_list(self):
+        tpl_summary = get_template_module('events/registration/display/_registration_summary_blocks.html')
+        return tpl_summary.render_receipts_list(self.registration, from_management=True)
+
+
+class RHDownloadReceipt(RHManageReceiptBase):
+    """Download a receipt file from a registration."""
+
+    def _process(self):
+        return self.receipt_file.file.send()
+
+
+class RHPublishReceipt(RHManageReceiptBase):
+    """Publish or unpublish a receipt from a registration."""
+
+    @use_kwargs({
+        'publish': fields.Bool(required=True)
+    }, location='query')
+    def _process(self, publish):
+        if self.receipt_file.is_published == publish:
+            return jsonify_data()
+        if not publish and request.method == 'POST':
+            self.receipt_file.is_published = False
+            flash(_("Document '{}' successfully unpublished").format(self.receipt_file.file.filename), 'success')
+            logger.info('Document %r from registration %r was unpublished by %r', self.receipt_file.file.filename,
+                        self.registration, session.user)
+            return jsonify_data(html=self._render_receipts_list())
```

Comment:
This is probably not the best way to do the unpublish action (and there's a CSRF error sometimes when publishing, which might be caused by this).
Should I maybe split this into 2 RHs (for publish and unpublish)? Since I don't really see how I can unpublish with the same form